### PR TITLE
ref(dynamic_sampling): Redesign of dynamic sampling rules [TET-566]

### PR DESCRIPTION
**Pull request by iambriccardo**
This PR aims at implementing a new set of abstractions that simplify the definition of dynamic sampling biases and most importantly standardizes their implementation.

# Usage Example
In order to motivate this change, I will explain how it will work if we want to add a new bias named `foo` which requires `sample_rate` and `bar` params.

The addition of a bias is split in two phases:
* Creation of the bias
* Combination of the bias

> In order to get the final rules you can just use the `generate_rules()` method from `dynamic_sampling/rules/base.py`.

## Creation of the bias
1. Define the rule type inside `dynamic_sampling/rules/utils.py`
```python
class RuleType(Enum):
      ...
      IGNORE_HEALTH_CHECKS_RULE = "ignoreHealthChecks"
      FOO_RULE = "fooRule"
      BOOST_KEY_TRANSACTIONS_RULE = "boostKeyTransactions"
      ...
```
3. Create a file named `foo_bias.py` inside `dynamic_sampling/rules/biases/`.
4. Create a data provider with `sample_rate` and `bar` params that extends `BiasDataProvider`
```python
class FooDataProvider(BiasDataProvider):
    def get_bias_data(self, bias_params: BiasParams) -> BiasData:
        return {
            "sampleRate": some_computation(),
            "bar": ["i", "like", "this"]
        }
```
5. Create a rule generator that uses the params from the data provider and that extends `BiasRulesGenerator`
```python
class FooDataRulesGenerator(BiasRulesGenerator):
    def _generate_bias_rules(self, bias_data: BiasData) -> List[BaseRule]:
        return [
            {
                "sampleRate": bias_data["sampleRate"],
                "type": "trace",
                "condition": {
                    "op": "or",
                    "inner": [
                        {
                            "op": "glob",
                            "name": "trace.environment",
                            "value": bias_data["bar"],
                            "options": {"ignoreCase": True},
                        }
                    ],
                },
                "active": True,
                "id": 1234,
            }
        ]
```
4. Define the bias by binding the data provider with the rules generator
```python
class FooBias(Bias):
    def __init__(self) -> None:
        super().__init__(FooDataProvider, FooRulesGenerator)
```

## Combination of the bias
1. Add the following line to `dynamic_sampling/rules/combine.py`
```python
...
def get_relay_biases_combinator():
    default_combinator = _get_default_combinator()

    ...
    default_combinator.combine(RuleType.FOO_RULE, FooBias()) # You can combine the rule in the position you prefer, based on the priority.
    ...
```

# Rationale
We are currently actively iterating on new biases and in the future we expect to add new ones in case new use-cases will emerge. For this reason it is important to build a solid skeleton that standardizes how we create biases.

The idea of standardizing biases comes from the fact that we saw how every implementation of biases was a little bit different and this difference is expected to be even more noticeable if a new more complex biases have to be implemented.

# Behind the Design
Before trying to understand the implementation a differentiation of bias and rule should be done:
* Bias -> a bias is the heuristic we use for dynamic sampling.
* Rule -> a rule is a formal definition of the bias that can be understood by Relay.

## Layers
The high-level design of this implementation revolves around a set of layers:
* **Data-bias-rules business logic** -> this part contains all the classes needed to fetch the data and generate the rule for each specific bias.
* **Biases combination** -> this part contains all the classes responsible of combining biases that generate rules.

### Data-bias-rules business logic
The base classes in this layer are:
* `BiasDataProvider` -> responsible of defining how to get required data for a bias.
* `BiasRulesGenerator` -> responsible of generating rules from a bias.
* `Bias` -> responsible of representing a bias and abstracting the data provider and rules generator.

The creation of a `Bias` can be done by combining different data providers and rules generator, however, in our implementation we hardcode the biases with the specific combination that are pre-defined. Technically a `Bias` is nothing more than a class that transparently ties together a `BiasDataProvider` and a `BiasRulesGenerator`.

### Biases combination
Because biases can be combined together, we designed an abstraction to generalize how they are combined, so that in the future we can very easily change how they are sorted in case Relay decides to change its matching heuristics.

The base classes in this layer are:
* `BiasesRulesCombinator` -> responsible of combining multiple biases with a given order dictated by the `order_discriminant`.

_This work is part of an effort to make the extension and evolution of dynamic sampling as seamless as possible._

### CODE REVIEW

File: src/sentry/dynamic_sampling/rules/base.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+                # If the customer has a base sample rate of 100% it means that we don't want to activate biases which
+                # means that dynamic sampling is not enabled (technically it is enabled, but logically it is
+                # indistinguishable from disabled).
+                return base_sample_rate < 1.0
+
+            # TODO: try to think of a more extensible implementation that guarantees deterministic ordering.
+            if RuleType.BOOST_KEY_TRANSACTIONS_RULE.value in enabled_biases:
+                combinator.combine_if(BOOST_KEY_TRANSACTIONS_BIAS, dynamic_sampling_enabled)
+
+            if RuleType.BOOST_ENVIRONMENTS_RULE.value in enabled_biases:
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/base.py:64]: Mapping ClassName: order

**Comment by matejminar** [comment]
That is a very useful description 👏
Let's copy-paste it into the notion under Telemetry Experience so that it can be easily referenced in the future.

**Comment by iambriccardo** [comment]
> That is a very useful description 👏 Let's copy-paste it into the notion under Telemetry Experience so that it can be easily referenced in the future.

Will do a refined explanation as soon as I have the approval from @andriisoldatenko.

File: src/sentry/api/endpoints/project_details.py [hunk 2]
```diff
@@ -23,8 +23,8 @@
 from sentry.auth.superuser import is_active_superuser
 from sentry.constants import RESERVED_PROJECT_SLUGS
 from sentry.datascrubbing import validate_pii_config_update
-from sentry.dynamic_sampling.rules_generator import generate_rules
-from sentry.dynamic_sampling.utils import get_supported_biases_ids, get_user_biases
+from sentry.dynamic_sampling.rules.base import generate_rules
```
Comment (andriisoldatenko) [inline src/sentry/api/endpoints/project_details.py:26]: you can declare these functions in dynamic_sampling `__init__.py` Similar
https://github.com/getsentry/sentry/blob/9af57a33c19004dd6d8ddb54cf55baa3d91146f6/src/sentry/receivers/__init__.py#L1-L16 and avoid such changes.

Comment (iambriccardo) [reply src/sentry/api/endpoints/project_details.py:26]: I actually did that in my first iterations but then had problems with mypy even after adding the comment.

Comment (iambriccardo) [reply src/sentry/api/endpoints/project_details.py:26]: I will try again to implement it and see which errors I get.

File: src/sentry/dynamic_sampling/rules/base.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+from sentry.models import Project
+
+ALWAYS_ALLOWED_RULE_TYPES = {RuleType.UNIFORM_RULE}
+
+
+def _get_guarded_blended_sample_rate(project: Project) -> float:
+    sample_rate = quotas.get_blended_sample_rate(project)
+
+    if sample_rate is None:
+        raise Exception("The method get_blended_sample_rate() returned None")
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/base.py:19]:
```suggestion
        raise Exception("The method get_blended_sample_rate() returns none")
```
or even better to stick with old message, so sentry can group it i guess.
```
get_blended_sample_rate returns none
```

File: src/sentry/dynamic_sampling/rules/combine.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+from sentry.dynamic_sampling.rules.utils import RuleType
+
+
+def _get_default_combinator() -> BiasesCombinator:
+    # The default combinator is the ordered combinator, which will keep the insertion order of the rules.
+    return OrderedBiasesCombinator()
+
+
+def get_relay_biases_combinator() -> BiasesCombinator:
+    default_combinator = _get_default_combinator()
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/combine.py:19]: i don't understand why do we need method to return an instance of the class? why not just replace with `default_combinator = OrderedBiasesCombinator()`

File: src/sentry/dynamic_sampling/rules/combine.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+
+    # The combination depends on the default_combinator used but in case of the ordered combinator the first combined
+    # rule will be the first rule in the output (e.g., UNIFORM_RULE will be the last).
+    #
+    # The ordering is very important, especially because relay performs matching following a FIFO matching algorithm.
+    #
+    # If you need to add any new bias, add it here after having created all the necessary classes.
+    default_combinator.combine(RuleType.BOOST_KEY_TRANSACTIONS_RULE, BoostKeyTransactionsBias())
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/combine.py:27]:
```suggestion
    default_combinator.add(RuleType.BOOST_KEY_TRANSACTIONS_RULE, BoostKeyTransactionsBias())
```
combinator.combine sounds a bit wordy for me.

File: src/sentry/dynamic_sampling/rules/combine.py (hunk 4 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/combine.py:19]: The reasoning was that I wanted to explicitly isolate the combinator definition but we can for sure inline it.

File: src/sentry/dynamic_sampling/rules/data/key_transactions.py [hunk 6]
```diff
@@ -1,7 +1,7 @@
 from typing import TYPE_CHECKING, List
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/data/key_transactions.py:1]: Entire `data` folder confuse me a lot:
- data sounds like data folder with data
- IMO it's just 1 file per func, so i prefer to move it to utils may be?
I understand the point you are trying to isolate later of data ... but TBH it do nothing, each bias has different data input, also we don't introduce any new abstract classes here. 🤔

File: src/sentry/dynamic_sampling/rules/combinators/base.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
+from abc import ABC, abstractmethod
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/combinators/base.py:1]: why do we have `combinators` module and separately `combine.py`?
Does it make sense to move `combine.py` into `combinators` folder?
Or even to `__init__.py`

File: src/sentry/dynamic_sampling/rules/base.py [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+from sentry.dynamic_sampling.rules.biases.base import Bias, BiasParams
+from sentry.dynamic_sampling.rules.combine import get_relay_biases_combinator
+from sentry.dynamic_sampling.rules.logging import log_rules
+from sentry.dynamic_sampling.rules.utils import BaseRule, RuleType, get_enabled_user_biases
+from sentry.models import Project
+
+ALWAYS_ALLOWED_RULE_TYPES = {RuleType.UNIFORM_RULE}
+
+
+def _get_guarded_blended_sample_rate(project: Project) -> float:
```
Comment (andriisoldatenko) [inline src/sentry/dynamic_sampling/rules/base.py:15]: i don't recall why it has `_` in the name.


Single underscore at the beginning:

Python doesn't have real private methods. Instead, one underscore at the start of a **method** or attribute name means you shouldn't access this method, because it's not part of the API.

So in this case you can just rename to `get_guarded_blended_sample_rate `

File: src/sentry/dynamic_sampling/rules/combinators/base.py (hunk 7 above)
Comment (iambriccardo) [reply src/sentry/dynamic_sampling/rules/combinators/base.py:1]: My idea was to keep `combine` at a high level because it didn't require much knowledge of the code structure to be used and it has as dependencies the combinators but nothing that extends them.

**Review by andriisoldatenko — APPROVED** [review]
